### PR TITLE
Impl range write

### LINKED ISSUE #360: Implement DW_AT_ranges writing
This is needed for https://github.com/bjorn3/rustc_codegen_cranelift/issues/166 line info, because I can't rely on faerie putting functions in a certain order. And gdb doesn't accept line info when neither `DW_AT_low_pc` nor `DW_AT_high_pc`/`DW_AT_ranges` is set.

### PR DESCRIPTION
**Pull request by bjorn3**
Fixes #360

<strike>I have only implemented DWARF 4 `.debug_ranges` writing, not DWARF 5 `.debug_rnglists` yet.</strike>

Edit: implemented `.debug_rnglists`, but didn't test it. (I have no DWARF 5 supporting dwarfdump)

### CODE REVIEW

**Comment by bjorn3** [comment]
It is failing for https://github.com/gimli-rs/gimli/pull/362/files#diff-b0573ccce68b7d366aa6a11719034ce8R1815.

```
thread 'write::unit::tests::test_attribute_value' panicked at 'assertion failed: `(left == right)`
  left: `Attribute { name: DwAt(85), value: RangeListsRef(RangeId(1)) }`,
 right: `Attribute { name: DwAt(85), value: RangeListsRef(RangeId(0)) }`', src/write/unit.rs:1958:25
```

**Comment by bjorn3** [comment]
I found the problem.

**Comment by bjorn3** [comment]
I have tested `.debug_ranges` with cg_clif and it seems to work correctly.

**Comment by philipc** [comment]
> Could you please test .debug_rnglists for me, as I don't have DWARF 5 aware tools?

Changing version to 5 and running dwarfdump on target/out/mini_core.dummy_name.rcgu.o looks ok. Note that gimli's dwarfdump should be able to read everything we write.

**Comment by bjorn3** [comment]
This is ready for review.

**Comment by bjorn3** [comment]
When enabling dwarf version 5 in cg_clif, I get `Failed to dump attribute value: The end of an address range must not be before the beginning.`:

```
dwarfdump target/out/mod_bench | grep -A 20 AT_range
                    DW_AT_ranges                0x0000000c
Failed to dump attribute value: The end of an address range must not be before the beginning.

LOCAL_SYMBOLS:
< 1><0x0000002a>    DW_TAG_subprogram
                      DW_AT_linkage_name          _ZN4core3num21_$LT$impl$u20$u32$GT$9max_value17hd2ebfb9c786040faE
                      DW_AT_name                  dummy_ZN4core3num21_$LT$impl$u20$u32$GT$9max_value17hd2ebfb9c786040faE
                      DW_AT_low_pc                0x000006c0
                      DW_AT_frame_base            len 0x0001: 56: DW_OP_reg6
                      DW_AT_decl_file             /home/bjorn/Documenten/rustc_codegen_cranelift/build_sysroot/sysroot_src/src/libcore/num/mod.rs
                      DW_AT_decl_line             0x00000888
                      DW_AT_decl_column           0x0000000c
                      DW_AT_high_pc               <offset-from-lowpc>39
< 2><0x0000004c>      DW_TAG_variable
                        DW_AT_name                  <no name>
                        DW_AT_type                  <0x00000059>
                        DW_AT_location              len 0x0002: 916c: DW_OP_fbreg -20
< 1><0x00000059>    DW_TAG_base_type
                      DW_AT_name                  u32
                      DW_AT_encoding              DW_ATE_unsigned
                      DW_AT_byte_size             0x00000004
--
                    DW_AT_ranges                0x00000216
                ranges: 1785 at .debug_rnglists offset 534 (0x00000216)
                        [ 0] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [ 1] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [ 2] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [ 3] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [ 4] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [ 5] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [ 6] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [ 7] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [ 8] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [ 9] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [10] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [11] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [12] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [13] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [14] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [15] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [16] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [17] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
                        [18] <start-end low-off: 0x00000000 addr 0x00000000 high-off: 0x00000000 addr 0x00000000>
```

File: src/write/range.rs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+#[derive(Debug, Default)]
+pub struct RangesTable {
+    ranges: IndexSet<RangeList>,
+}
+
+impl RangesTable {
+    /// Add an range to the table and return its code.
+    pub fn add(&mut self, range_list: RangeList) -> RangeId {
+        let (code, _) = self.ranges.insert_full(range_list);
+        // Code must be non-zero
```
Comment (philipc) [inline src/write/range.rs]:
```suggestion
```

File: src/write/range.rs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        RangeId(code)
+    }
+
+    /// Write the range table to the appropriate section for the given DWARF version
```
Comment (philipc) [inline src/write/range.rs]:
```suggestion
    /// Write the range table to the appropriate section for the given DWARF version.
```

File: src/write/range.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    pub fn write<W: Writer>(
+        &self,
+        w_ranges: &mut DebugRanges<W>,
+        w_rnglists: &mut DebugRngLists<W>,
+        format: Format,
+        version: u16,
+        address_size: u8,
+    ) -> Result<(DebugRangesOffsets, DebugRngListsOffsets)> {
+        if version < 5 {
```
Comment (philipc) [inline src/write/range.rs]: Please make the version check more restrictive, such as in: https://github.com/gimli-rs/gimli/blob/f3116ec8654be37c0aa281d7656294b21fd05adc/src/write/unit.rs#L236-L253.

File: src/write/range.rs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            Ok((self.write_ranges(w_ranges, address_size)?, DebugRngListsOffsets::default()))
+        } else {
+            Ok((DebugRangesOffsets::default(), self.write_rnglists(w_rnglists, format, version, address_size)?))
+        }
+    }
+
+    /// Write the range table to the `.debug_ranges` section.
+    pub fn write_ranges<W: Writer>(&self, w: &mut DebugRanges<W>, address_size: u8) -> Result<DebugRangesOffsets> {
```
Comment (philipc) [inline src/write/range.rs]: I don't think `write_ranges` or `write_rnglists` need to be public?

File: src/write/range.rs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    DebugRangesOffsets: RangeId => RangeListsOffset,
+    "The section offsets of all ranges within a `.debug_ranges` section."
+);
+
+define_offsets!(
+    DebugRngListsOffsets: RangeId => RangeListsOffset,
+    "The section offsets of all ranges within a `.debug_rnglists` section."
+);
+
+/// An identifier for a unit in a `UnitTable`.
```
Comment (philipc) [inline src/write/range.rs]: Fix comment.

File: src/write/range.rs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+#[cfg(feature = "read")]
+mod convert {
+    use super::*;
+
+    use read::{self, Reader};
+    use write::ConvertResult;
+    impl RangeList {
+        /// Create an attribute by reading the data in the given sections.
```
Comment (philipc) [inline src/write/range.rs]: Fix comment.

File: src/write/range.rs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        ) -> ConvertResult<Option<Self>> {
+            let mut ranges = Vec::new();
+            while let Some(range) = from.next()? {
+                if let (Some(begin), Some(end)) = (convert_address(range.begin), convert_address(range.end)) {
+                    ranges.push(Range {
+                        begin,
+                        end,
+                    });
+                } else {
+                    return Ok(None)
```
Comment (philipc) [inline src/write/range.rs]:
```suggestion
                    return Err(ConvertError::InvalidAddress);
```

File: src/write/unit.rs [hunk 8]
```diff
@@ -635,7 +649,7 @@ pub enum AttributeValue {
     /// It is the user's responsibility to ensure the offset is valid.
     /// This variant will be removed from the API once support for writing
     /// `.debug_ranges`/`.debug_rnglists` sections is implemented.
-    RangeListsRef(RangeListsOffset),
+    RangeListsRef(RangeId),
```
Comment (philipc) [inline src/write/unit.rs]:
```suggestion
    RangeListRef(RangeListId),
```

File: src/write/unit.rs [hunk 9]
```diff
@@ -635,7 +649,7 @@ pub enum AttributeValue {
     /// It is the user's responsibility to ensure the offset is valid.
     /// This variant will be removed from the API once support for writing
     /// `.debug_ranges`/`.debug_rnglists` sections is implemented.
```
Comment (philipc) [inline src/write/unit.rs]: Fix comment.

File: src/write/range.rs [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+);
+
+define_offsets!(
+    DebugRngListsOffsets: RangeId => RangeListsOffset,
+    "The section offsets of all ranges within a `.debug_rnglists` section."
+);
+
+/// An identifier for a unit in a `UnitTable`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct RangeId(pub usize);
```
Comment (philipc) [inline src/write/range.rs]: Rename to `RangeListId`.

File: src/write/range.rs [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+/// A table of ranges that will be stored in a `.debug_ranges` or `.debug_rnglists` section.
+#[derive(Debug, Default)]
+pub struct RangesTable {
```
Comment (philipc) [inline src/write/range.rs]: Rename to `RangeListTable`.

File: src/write/range.rs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+    /// Write the range table to the appropriate section for the given DWARF version
+    pub fn write<W: Writer>(
+        &self,
+        w_ranges: &mut DebugRanges<W>,
+        w_rnglists: &mut DebugRngLists<W>,
+        format: Format,
+        version: u16,
+        address_size: u8,
+    ) -> Result<(DebugRangesOffsets, DebugRngListsOffsets)> {
```
Comment (philipc) [inline src/write/range.rs]: Maybe we should return a struct:
```rust
struct RangeListOffsets {
   debug_ranges: DebugRangesOffsets,
   debug_rnglists: DebugRngListsOffsets,
}
```
That'll mean one less parameter that needs to be passed around elsewhere.

File: src/write/unit.rs [hunk 13]
```diff
@@ -1317,18 +1338,35 @@ mod convert {
...
                 }
                 read::AttributeValue::DebugRngListsIndex(index) => {
-                    let offset = context
+                    let _offset = context
                         .dwarf
                         .ranges
                         .get_offset(context.rnglists_base, index)?;
-                    AttributeValue::RangeListsRef(offset)
+                    //AttributeValue::RangeListsRef(offset)
+                    unimplemented!();
```
Comment (philipc) [inline src/write/unit.rs]: Please implement. This should be as simple as passing the offset to the same code that handles `read::AttributeValue::RangeListsRef`.

**Comment by philipc** [comment]
This won't support conversion of relocatable range lists. For that we'll need to switch to converting the raw ranges, but I'm happy to leave that until another PR.

Normally I would also want more tests in `write::range`, but those tests will need significant changes for the raw range support, so it's okay to skip that for now.

File: src/write/range.rs [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+);
+
+/// An identifier for a unit in a `UnitTable`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct RangeId(pub usize);
+
+/// A table of ranges that will be stored in a `.debug_ranges` or `.debug_rnglists` section.
+#[derive(Debug, Default)]
+pub struct RangesTable {
+    ranges: IndexSet<RangeList>,
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply src/write/range.rs]: You mean to improve performance?

[the remark this answers is not in the document]
Comment (philipc) [reply src/write/range.rs]: Yes. Probably won't make much difference to performance in the whole either. Mostly I just want to document that use of `IndexSet` probably isn't strictly required.

**Comment by philipc** [comment]
This is looking good. Can you please:
- fix the failing test (cleanest way is probably to move it to a separate test?)
- run rustfmt
- merge into logical commits (one commit is probably fine)

File: src/write/range.rs [hunk 15]
```diff
@@ -54,16 +54,16 @@ impl RangeListTable {
         version: u16,
         address_size: u8,
     ) -> Result<RangeListOffsets> {
-        if version < 5 {
-            Ok(RangeListOffsets {
+        match version {
+            3...4 => Ok(RangeListOffsets {
```
Comment (philipc) [inline src/write/range.rs]: Both gcc and llvm write `.debug_ranges` even for DWARF version 2, so we should do the same.

**Comment by bjorn3** [comment]
Done

**Comment by philipc** [comment]
Thanks!

**Pull request merged**